pkg/controller/servicebindingrequest/envvars/envvars.go

```diff
@@ -0,0 +1,128 @@
+package envvars
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/imdario/mergo"
+)
+
+// ErrUnsupportedType is returned when an unsupported type is encountered.
+var ErrUnsupportedType = errors.New("unsupported type")
+
+// Build returns an environment variable dictionary with an entry for each
+// leaf containing a scalar value.
+//
+// path indicates the node that should be treated as root, and informing an empty slice
+//
+// Example:
+//
+// 	src := map[string]interface{}{
+// 		"status": map[string]interface{}{
+//			"listeners": []map[string]interface{}{
+//				{
+//					"type": "secure",
+//					"addresses": []map[string]interface{}{
+//						{
+//							"host": "my-cluster-kafka-bootstrap.coffeeshop.svc",
+//							"port": "9093",
+//						},
+//					},
+//				},
+//			},
+//		},
+//	}
+//	actual, _ := Build(src, []string{})
+//
+// actual should contain the following values:
+//
+// 	"STATUS_LISTENERS_0_TYPE":             "secure",
+// 	"STATUS_LISTENERS_0_ADDRESSES_0_HOST": "my-cluster-kafka-bootstrap.coffeeshop.svc",
+// 	"STATUS_LISTENERS_0_ADDRESSES_0_PORT": "9093",
+//
+// Now, consider the following example:
+//
+//	actual, _ = Build(src, []string{"status"})
+//
+// actual should contain the following values instead:
+//
+// 	"LISTENERS_0_TYPE":             "secure",
+// 	"LISTENERS_0_ADDRESSES_0_HOST": "my-cluster-kafka-bootstrap.coffeeshop.svc",
+// 	"LISTENERS_0_ADDRESSES_0_PORT": "9093",
+//
+func Build(obj interface{}, path []string) (map[string]string, error) {
```

Comment:
As the above example demonstrates, and if I can see the usage in the code, public API expect that `obj` should be actually a map, right? Thus let's restrict the public api:

```
func Build(obj map[string]interface{}, path []string) (map[string]string, error) {
   return doBuild(obj, path)
}

func doBuild(obj interface{}, path []string) (map[string]string, error) {
// old Build function
}
```

all other internal functions should use `doBuild` instead.

Reply:
Please submit this change as a PR if you'd like to proceed with that.